### PR TITLE
Rspec

### PR DESCRIPTION
**Pull request by deikdesign**
## Add Rspec tests  to functions in module
- add test cases for : my_each, my_each_with_index, my_select, my_all, my_any, my_none, my_count, my_map, my_inject

### CODE REVIEW

File: spec/methods_spec.rb [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+describe Enumerable do
+let(:array) { [1, 2, 3, 4, 5]}
+let(:range) { (0..5) }
+let(:hash)  { { a: 1, b: 2, c: 3} }
+let(:my_proc) { proc { |num| num + 1} }
+
+    describe '#my_each' do
+         it 'return correct array'
+         expect(array.my_each { |num| num > 2}).to eql(array)
+          end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** You have not added `do` after starting `it` block. Please fix it.

File: spec/methods_spec.rb [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        expect(hash.my_all? { |n| n > 2 }).to eql FALSE
+        end
+
+        it 'return true when all values are true in a range' do
+        expect(hash.my_all? { |n| n > 6 }).to eql TRUE
+        end
+
+        it 'return true when all values are true in a hash' do
+        expect(hash.my_all? { |key, value| value > 7 }).to eql TRUE
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** You have not added `end` after finishing `it` block. Please fix it.

- [x] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_all return true when all values are true in a hash
     Failure/Error: expect(hash.my_all? { |key, value| value > 7 }).to eql TRUE
     
       expected: true
            got: false
     
       (compared using eql?)
     
       Diff:
       @@ -1 +1 @@
       -true
       +false
       
     # ./spec/methods_spec.rb:74:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return false when one value is false in a range' do
+        expect(hash.my_all? { |key, value| value > 3 }).to eql TRUE
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** You have not added `end` after finishing `it` block. Please fix it.

- [x] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_all return false when one value is false in a range
     Failure/Error: expect(hash.my_all? { |key, value| value > 3 }).to eql TRUE
     
       expected: true
            got: false
     
       (compared using eql?)
     
       Diff:
       @@ -1 +1 @@
       -true
       +false
       
     # ./spec/methods_spec.rb:78:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    end
+
+    describe '#my_none?' do
+        it 'return true when no value is true on an array' do
+        expect(range.my_none? { |n| n > 6}  ).to eql TRUE
+        end
+
+        it 'return true when any value is true on an array' do
+        expect(range.my_none? { |n| n > 2}  ).to eql FALSE
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** You have not added `end` after finishing `it` block. Please fix it.

File: spec/methods_spec.rb [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    end    
+
+    describe '#my_count?' do
+        it 'return to an array when no block  and no argument are given' do
+        expect(array.my_count?).to eql(5)
+        end
+
+        it 'elements in a range when no block  and no argument are given' do
+        expect(range.my_count?).to eql(6)
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** You have not added `end` after finishing `it` block. Please fix it.

- [x] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_count? return to an array when no block  and no argument are given
      Failure/Error: expect(array.my_count?).to eql(5)
      
      NoMethodError:
        undefined method `my_count?' for [1, 2, 3, 4, 5]:Array
        Did you mean?  my_count
      # ./spec/methods_spec.rb:136:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+         end 
+          it 'when block is not given return Enumerator' do
+          expect(array.my_each).to be_an(Enumerator)
+         end
+    end
+
+    describe '#my_each_with_index' do
+          it 'return correct array' do
+          expect(range.my_each_with_index { |num, index| num}).to eql(hash)
+          end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_each_with_index return correct array
     Failure/Error: expect(range.my_each_with_index { |num, index| num}).to eql(hash)
     
       expected: {:a=>1, :b=>2, :c=>3}
            got: 0..5
     
       (compared using eql?)
     
       Diff:
       @@ -1,4 +1,2 @@
       -:a => 1,
       -:b => 2,
       -:c => 3,
       +0..5
       
     # ./spec/methods_spec.rb:26:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+         end
+
+        it 'returns to an array' do
+        expect(array.my_select { |num| num > 3}).to eql([4, 5])
+        end
+
+        it 'returns an array into a range' do
+        expect(array.my_select { |num| num < 3}).to eql([0, 1, 2])
+        end
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_select returns an array into a range
     Failure/Error: expect(array.my_select { |num| num < 3}).to eql([0, 1, 2])
     
       expected: [0, 1, 2]
            got: [1, 2]
     
       (compared using eql?)
     # ./spec/methods_spec.rb:49:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        it 'returns an array into a hash' do
+        expect(array.my_select { |key, value| value < 2}).to eql({ c: 3})
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
 Enumerable#my_select returns an array into a hash
     Failure/Error: expect(array.my_select { |key, value| value < 2}).to eql({ c: 3})
     
     NoMethodError:
       undefined method `<' for nil:NilClass
     # ./spec/methods_spec.rb:53:in `block (4 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'if all value are false for an array return false' do
+        expect(array.my_all? { |n| n > 6}).to eql FALSE
+        end
+
+        it 'return false when one value is false in a range' do
+        expect(hash.my_all? { |n| n > 2 }).to eql FALSE
+        end
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_all return false when one value is false in a range
     Failure/Error: expect(hash.my_all? { |n| n > 2 }).to eql FALSE
     
     NoMethodError:
       undefined method `>' for [:a, 1]:Array
     # ./spec/methods_spec.rb:66:in `block (4 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return true when all values are true in a range' do
+        expect(hash.my_all? { |n| n > 6 }).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
 Enumerable#my_all return true when all values are true in a range
     Failure/Error: expect(hash.my_all? { |n| n > 6 }).to eql TRUE
     
     NoMethodError:
       undefined method `>' for [:a, 1]:Array
     # ./spec/methods_spec.rb:70:in `block (4 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        expect(range.my_any? { |n| n > 2}  ).to eql TRUE
+        end
+
+        it 'return false when no value is true on the range' do
+        expect(range.my_any? { |n| n > 6}  ).to eql FALSE
+        end
+       
+        it 'return false when any value is true on a hash' do
+        expect(range.my_any? { |key, value| value > 2}  ).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_any? return false when any value is true on a hash
     Failure/Error: expect(range.my_any? { |key, value| value > 2}  ).to eql TRUE
     
     NoMethodError:
       undefined method `>' for nil:NilClass
     # ./spec/methods_spec.rb:99:in `block (4 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return false when no value is true on a hash' do
+        expect(range.my_any? { |key, value| value > 6}  ).to eql false
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_any? return false when no value is true on a hash
     Failure/Error: expect(range.my_any? { |key, value| value > 6}  ).to eql false
     
     NoMethodError:
       undefined method `>' for nil:NilClass
     # ./spec/methods_spec.rb:103:in `block (4 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return to an array when no block  and no argument are given' do
+        expect(array.my_count?).to eql(5)
+        end
+
+        it 'elements in a range when no block  and no argument are given' do
+        expect(range.my_count?).to eql(6)
+
+        it 'elements in a hash when no block  and no argument are given' do
+        expect(hash.my_count?).to eql(3)
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
 Enumerable#my_count? elements in a range when no block  and no argument are given
      Failure/Error: expect(range.my_count?).to eql(6)
      
      NoMethodError:
        undefined method `my_count?' for 0..5:Range
        Did you mean?  my_count
      # ./spec/methods_spec.rb:140:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return elements when block is given' do
+        expect(array.my_count? { |x| x == 2}  ).to eql(1)
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_count? elements in a hash when no block  and no argument are given
      Failure/Error: expect(hash.my_count?).to eql(3)
      
      NoMethodError:
        undefined method `my_count?' for {:a=>1, :b=>2, :c=>3}:Hash
        Did you mean?  my_count
      # ./spec/methods_spec.rb:144:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return elements are equal to arg value' do
+        expect(array.my_count(1)).to eql(1)
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_count? return elements when block is given
      Failure/Error: expect(array.my_count? { |x| x == 2}  ).to eql(1)
      
      NoMethodError:
        undefined method `my_count?' for [1, 2, 3, 4, 5]:Array
        Did you mean?  my_count
      # ./spec/methods_spec.rb:148:in `block (3 levels) in <top (required)>'
```

File: spec/methods_spec.rb [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+        expect(array.my_inject { |sum, num| sum + num } ).to eql 15
+        end
+
+        it 'inject range if the block is given' do
+        expect(range.my_inject { |sum, num| sum + num } ).to eql 15
+        end
+    
+        it 'inject enumerator if arg is given' do
+        expect(range.my_inject).to be_an(Enumerator)
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Please fix this test:
```ruby
Enumerable#my_inject? inject enumerator if arg is given
      Failure/Error: to_a.my_each { |item| int = int.nil? ? item : yield(int, item) }
      
      LocalJumpError:
        no block given (yield)
```

File: spec/methods_spec.rb [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'returns an array into a hash' do
+        expect(array.my_select { |key, value| value < 2}).to eql({ c: 3})
+        end
+    end
+    describe '#my_all' do
+        it 'if all value for an array are true return true' do
+        expect(array.my_all? { |n| n > 2}).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'if all value are false for an array return false' do
+        expect(array.my_all? { |n| n > 6}).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return false when one value is false in a range' do
+        expect(hash.my_all? { |n| n > 2 }).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        expect(hash.my_all? { |key, value| value > 3 }).to eql TRUE
+    end
+    describe '#my_any?' do
+        it 'return true when any value is true on the array' do
+        expect(array.my_any? { |n| n > 2}).to eql TRUE
+        end
+
+        it 'return false when no value is true on the array' do
+        expect(array.my_any? { |n| n > 6}).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return true when any value is true on the range' do
+        expect(range.my_any? { |n| n > 2}  ).to eql TRUE
+        end
+
+        it 'return false when no value is true on the range' do
+        expect(range.my_any? { |n| n > 6}  ).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb (hunk 4 above)
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return true when any value is true on an array' do
+        expect(range.my_none? { |n| n > 2}  ).to eql FALSE
+
+        it 'return false when any value is true on a range' do
+        expect(range.my_none? { |n| n > 6}  ).to eql TRUE
+        end
+
+        it 'return false when any value is false on a range' do
+        expect(range.my_none? { |n| n > 2}  ).to eql FALSE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `FALSE` is deprecated. Please use the small case `false` instead.

File: spec/methods_spec.rb (hunk 10 above)
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return false when one value is false in a range' do
+        expect(hash.my_all? { |n| n > 2 }).to eql FALSE
+        end
+
+        it 'return true when all values are true in a range' do
+        expect(hash.my_all? { |n| n > 6 }).to eql TRUE
+        end
+
+        it 'return true when all values are true in a hash' do
+        expect(hash.my_all? { |key, value| value > 7 }).to eql TRUE
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb (hunk 3 above)
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return false when one value is false in a range' do
+        expect(hash.my_all? { |key, value| value > 3 }).to eql TRUE
+    end
+    describe '#my_any?' do
+        it 'return true when any value is true on the array' do
+        expect(array.my_any? { |n| n > 2}).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return false when no value is true on the array' do
+        expect(array.my_any? { |n| n > 6}).to eql FALSE
+        end
+
+        it 'return true when any value is true on the range' do
+        expect(range.my_any? { |n| n > 2}  ).to eql TRUE
+        end
+
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb (hunk 11 above)
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+        it 'return false when no value is true on a hash' do
+        expect(range.my_any? { |key, value| value > 6}  ).to eql false
+        end
+    end
+
+    describe '#my_none?' do
+        it 'return true when no value is true on an array' do
+        expect(range.my_none? { |n| n > 6}  ).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [ ] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+
+        it 'return true when any value is true on an array' do
+        expect(range.my_none? { |n| n > 2}  ).to eql FALSE
+
+        it 'return false when any value is true on a range' do
+        expect(range.my_none? { |n| n > 6}  ).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

File: spec/methods_spec.rb [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+        expect(range.my_none? { |n| n > 2}  ).to eql FALSE
+        end
+
+        it 'return true when no value is true on a range' do
+        expect(range.my_none? { |key, value| key > 6}  ).to eql TRUE
+        end
+
+        it 'return true when all values are false on the hash' do
+        expect(range.my_none? { |key, value| key > 6}  ).to eql TRUE
+        end
```
Comment (moinkhanif) [inline spec/methods_spec.rb]: - [x] **[REQUIRED]** Use of All Capital `TRUE` is deprecated. Please use the small case `true` instead.

**Review by moinkhanif — CHANGES_REQUESTED** [review]
## Changes Required❗

Hey @deikdesign , thanks for submitting the project. However, there are a few **missing `do` or `end` keywords** that are causing issues. Also, **14 of your tests fail**. Please make sure that all the **tests pass** before submitting a request for review.

Kindly also **fix linter errors** in the pull request. Make sure to **add a pull request description** as well.

I have mentioned the **rest of the required changes in the below comments**. Please fix them

**Happy Coding**

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle: 
### Hello Team! 
 
### You have done well so far :clap: However, there are some changes you have to make for your project to be approved. Kindly work on them and submit them for another code review.


 **General**



- [x] Your project looks great. However, your project cannot be approved until you write a descriptive Pull Request Message. It's essential that your repository is professional, as this is one of the things that potential colleagues will look at when recruiting team members. Please refer to [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243642/) for more information.

**Recommended**

##### PLEASE FIX THESE ERRORS
- [x]  Enumerable#my_any? return false when any value is true on a hash
     Failure/Error: expect(hash.my_any? { |n| n > 2 }).to eql true
          NoMethodError: undefined method `>' for [:a, 1]:Array 

- [x] Enumerable#my_any? return false when no value is true on a hash
     Failure/Error: expect(hash.my_any? { |n| n > 6 }).to eql false  
     NoMethodError:  undefined method `>' for [:a, 1]:Array


- [x] Enumerable#my_inject? inject enumerator if arg is given
     Failure/Error: to_a.my_each { |item| int = int.nil? ? item : yield(int, item) }
          LocalJumpError:  no block given (yield)


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review @tamara._ :wink: 
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by mahmoud717 — APPROVED** [review]
## Project approved ✔️  # 

Nice job implementing the required changes, you can go ahead and merge this pull request.

**Pull request merged**